File: go-problems/src/go-problems/problems/bash-command.go
/* 
RedHat Coding question
The following bash pipeline determines the frequency with which different words occur in a
given input.  It outputs the most frequent words, along with their frequencies.
    tr -s ' ' '\n' | sort | uniq -c | sort -n | tail

An example invocation is as follows (of course, on your computer the exact results may vary):
$ cat /usr/share/doc/coreutils-common/README | tr -s ' ' '\n' | sort | uniq -c | sort -n | tail

15 is
18 as
23 that

Implement an equivalent in Go.  You should use only the Go standard library, and you must not
use the os/exec package.  For a varied set of potential inputs, the output should be as close as
reasonable to that of the bash pipeline.

*/
package main


import (
    "fmt"
    "os"
	"strings"
	"sort"
	"io/ioutil"
)


func checkError(e error) {
	if e != nil {
		panic(e)
	}
}

func repetition(lines []string) map[string]int {
  
    wc := make(map[string]int)
	for i := 0; i < len(lines); i++ {

		word := strings.ReplaceAll(lines[i], " ", "")

		if word == "" {
			continue
		}
		
		word1 := strings.TrimSpace(word)
		_, matched := wc[word1]
        if matched {
            wc[word1] += 1
        } else {
            wc[word1] = 1
        }
	}
  
    return wc
}

func sortMapByValue(counts map[string]int) (map[string]int, []string){
	keys := make([]string, 0, len(counts))
    for key := range counts {
        keys = append(keys, key)
    }
    sort.Slice(keys, func(i, j int) bool { return counts[keys[i]] > counts[keys[j]] })
	return counts, keys
}


func main() {
	argsWithoutProg := os.Args[1:]
	// fmt.Println(argsWithoutProg)

	file_data, err := ioutil.ReadFile(argsWithoutProg[0])
	checkError(err)
	
	lines := strings.Split(string(file_data), " ")
	countsData := repetition(lines)

	
	sortedCountData, keys := sortMapByValue(countsData)

    for index, key := range keys {
		if sortedCountData[key] == 1 {
			continue
		}
		if index <= 10 {
			fmt.Printf("   %d %s\n", sortedCountData[key], key)
		}
    }
}

File: python-problems/bash_command.py
#!/usr/bin/python
"""

RedHat Coding question
The following bash pipeline determines the frequency with which different words occur in a
given input.  It outputs the most frequent words, along with their frequencies.
    tr -s ' ' '\n' | sort | uniq -c | sort -n | tail
An example invocation is as follows (of course, on your computer the exact results may vary):
$ cat /usr/share/doc/coreutils-common/README | tr -s ' ' '\n' | sort | uniq -c | sort -n | tail
15 is
18 as
23 that
Implement an equivalent in Python.  You should use only the python standard library, and you must not
use the exec library.  For a varied set of potential inputs, the output should be as close as
reasonable to that of the bash pipeline.

"""

import sys, os
import operator


def implement_bash_command(file_name):
    
    try:
        with open(_arg, 'r') as _file:
            d = dict()
            for line_data in _file.readlines():
                # Loop through each line of the file
                for line_data_list in line_data.splitlines():
                    for line in line_data_list.split(" "):
                        # Remove the leading spaces and newline character
                        line = line.rstrip()
                
                        # Convert the characters in line to 
                        # lowercase to avoid case mismatch
                        lines = line.lower()

                        # Split the line into words
                        words = line.split("\n")

                        # Iterate over each word in line
                        for word in words:
                            # Check if the word is already in dictionary
                            if word == '':
                                continue
                            word_lower = word
                            if word_lower in d:
                                # Increment count of word by 1
                                d[word_lower] = d[word_lower] + 1
                            else:
                                # Add the word to dictionary with count 1
                                d[word_lower] = 1
            
            # Print the contents of dictionary
            for key, _ in sorted(d.items(), key=operator.itemgetter(1), reverse=True)[:10]:
                if d[key] > 1:
                    print(" "*3 + str(d[key]), key)
                
    except IOError as _err:
        sys.stderr.write(os.path.basename(sys.argv[0]) + ": " + _arg + ": " + _err.strerror + "\n")

if __name__ == "__main__":
    for _arg in sys.argv[1:]:
       implement_bash_command(_arg)
